crates/engine/tree/src/tree/mod.rs

```diff
@@ -342,9 +366,38 @@ where
                     }
                 }
             }
+
+            if self.should_persist() {
+                self.persist_state().await;
+            }
         }
     }
 
+    /// Returns true if the canonical chain length minus the last persisted
+    /// block is more than the persistence threshold.
+    fn should_persist(&self) -> bool {
```

Comment:
I've started working on some tests for these methods, will add them in a follow up